### PR TITLE
Add reset gate

### PR DESCRIPTION
**Pull request by AthenaCaesura**
## Description

This PR adds the ResetOperation and the SX gate for use in resource estimation and instance generation.
It also includes a fix to `graph_sim_mini` which more cleverly chooses the method by which S gates are implemented.
It also improves the tests to `graph_sim_mini`.

I would request a particularly scrupulus review of `_qsp.py` as it seems that the number of qubits in circuits generated in this way has decreased. (in example 4 it has decreased from 25 to 15) Getting some explanation here would help a lot.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

File: src/benchq/compilation/graph_sim_mini.jl [hunk 1]
```diff
@@ -278,24 +290,35 @@ function get_icm(circuit, n_qubits::Int, with_measurements::Bool=false)
...
+        if occursin("ResetOperation", pyconvert(String, op.__str__()))
             original_qubit = get_qubit_1(op)
             compiled_qubit = qubit_map[original_qubit+1]
             qubit_map[original_qubit+1] = new_qubit = curr_qubits
             curr_qubits += 1
 
-            push!(compiled_circuit, ICMOp(CNOT_code, compiled_qubit, new_qubit))
             with_measurements &&
                 push!(compiled_circuit,
                       ICMOp(code_list[op_index]+MEASURE_OFFSET, compiled_qubit, new_qubit))
```
Comment (max-radin) [inline src/benchq/compilation/graph_sim_mini.jl:301]: Delete this lines since we don't need the measurement result.

**Review by max-radin — CHANGES_REQUESTED** [review]
Left some comments. Also it looks like the `defs.py` in example 6 still needs to have `simplify_rotations` updated to `transpile_to_native_gates`.

File: src/benchq/compilation/transpile_to_native_gates.py [hunk 2]
```diff
@@ -19,43 +21,37 @@
...
-    def production(self, operation: GateOperation) -> Iterable[GateOperation]:
-        q_index = operation.qubit_indices[0]
-        return [H(q_index), T(q_index), S(q_index), H(q_index), T(q_index), H(q_index)]
-
-
 class DecomposeStandardRZ(DecompositionRule[GateOperation]):
     """Decomposes RZ with characteristic angles to regular gates."""
 
     def predicate(self, operation: GateOperation) -> bool:
+        if not isinstance(operation, GateOperation):
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/compilation/transpile_to_native_gates.py:44]: Nice! I have added the necessary code and deleted the checks for `GateOperation` on each predicate.

File: src/benchq/compilation/transpile_to_native_gates.py [hunk 3]
```diff
@@ -144,3 +151,91 @@ def preprocess_gate(gate):
...
+        gate_operation_decomposition = [
+            preprocess_gate(gate)(*operation.qubit_indices)
+            for gate in gate_decomposition
+        ]
+
+        return reversed(gate_operation_decomposition)
+
+
+class CCZtoT(DecompositionRule[GateOperation]):
+    """Decomposition of U3 into RX and RZ gates."""
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/compilation/transpile_to_native_gates.py]: Fixed!

File: tests/benchq/compilation/test_transpilation.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 import pytest
 from orquestra.quantum.circuits import RX, RY, RZ, Circuit, Dagger, H, I, S, T, Z
 
-from benchq.compilation import simplify_rotations
+from benchq.compilation import transpile_to_native_gates
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply tests/benchq/compilation/test_transpilation.py]: done!

File: src/benchq/compilation/transpile_to_native_gates.py [hunk 5]
```diff
@@ -144,3 +151,91 @@ def preprocess_gate(gate):
...
+            T(qubit_2),
+            T(qubit_3),
+            H(qubit_3),
+            CNOT(qubit_1, qubit_2),
+            T.dagger(qubit_2),
+            T(qubit_1),
+            CNOT(qubit_1, qubit_2),
+        ]
+
+        return reversed(gate_operation_decomposition)
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/compilation/transpile_to_native_gates.py]: You're right! Sorry about the confusion!

File: src/benchq/data_structures/quantum_program.py [hunk 6]
```diff
@@ -74,7 +74,9 @@ def n_nodes(self) -> int:
     def count_gates_in_subroutine(self, step: int, gates: Sequence[str]) -> int:
         n_gates = 0
         for op in self.subroutines[step].operations:
-            if op.gate.name in gates:
+            if isinstance(op, GateOperation) and op.gate.name in gates:
+                n_gates += 1
+            if isinstance(op, ResetOperation) and "ResetOperation" in gates:
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/data_structures/quantum_program.py:79]: I would want to count them in the future. So I'll change the name to `count_operations` function. It's a pretty easy fix.

File: src/benchq/compilation/graph_sim_mini.jl [hunk 7]
```diff
@@ -145,21 +148,34 @@ function cz(lco, adj, vertex_1, vertex_2)
     end
 end
 
-function get_neighbor(neighbors, avoid)
+"""
+Select a neighbor to use when removing an LCO
+
+vb will be set to avoid if there are no neighbors, or avoid is the only neighbor,
```
[the remark this answers is not in the document]
Comment (AthenaCaesura) [reply src/benchq/compilation/graph_sim_mini.jl]: Sounds good! I renamed `vb` to `neighbor` just to make it clear what that variable is.

File: src/benchq/compilation/graph_sim_mini.jl (hunk 1 above)
Comment (AthenaCaesura) [reply src/benchq/compilation/graph_sim_mini.jl]: done!

**Review by max-radin — APPROVED** [review]

**Pull request merged**